add description and examples for `$anchor` keyword (#113)
Intelligence-AI/learnjsonschema.com: content/2020-12/core/anchor.markdown

Reviewer: I think another point worth mentioning is that defining two equal anchors on the SAME schema resource is considered to be an error.

However, defining two equal anchors on DIFFERENT schema resources (i.e. subschemas with their own `$id`) is valid. Maybe show an example of that?

Author: You mean something like this? ---
Valid:
```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/base",
  "properties": {
    "foo": {
      "$id": "one",
      "$anchor": "bar",
      "type": "integer"
    },
    "bar": {
      "$id": "two",
      "$anchor": "bar",
      "type": "string"
    }
  }
}
```

Invalid: 
```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/base",
  "properties": {
    "foo": {
      "$anchor": "bar",
      "type": "integer"
    },
    "bar": {
      "$anchor": "bar",
      "type": "string"
    }
  }
}
```

However, the implementations do not throw any error for the second schema and the fragmented URI reference is resolved to the last `$anchor`

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/base",
  "properties": {
    "foo": {
      "$anchor": "bar",
      "type": "integer"
    },
    "bar": {
      "$anchor": "bar",
      "type": "string"
    },
    "baz": {
      "$ref": "#bar"
    }
  }
```
Instance: `{"foo": 22, "bar": "aaaa", "baz": "aaaa" }`

Reviewer: > However, the implementations do not throw any error for the second schema and the fragmented URI reference is resolved to the last $anchor

That's strange. Which implementations are you testing with?

Reviewer: @jdesrosiers Shouldn't Hyperjump fail for the latter schema? i.e. same anchor twice in the same schema resource?

Reviewer 2: Having duplicate anchors doesn't make the schema invalid. It's just nonsense to do so and the behavior is undefined. It's like having a JSON value with duplicate properties (`{ "a": 1, "a": 2 }`). It's not invalid JSON, but you can't count on different implementations handling it the same way.

So, definitely tell users not to do that, but don't tell them they should expect an error.

Reviewer: Ah, interesting. I thought it would result in an error. Then probably fine to ignore all this @AgniveshChaubey

  - vocabulary: core
    keyword: $dynamicAnchor
---

> 

Reviewer: Extra blank line here?

> The `$anchor` keyword is used to assign a unique identifier to a subschema within the scope of its containing schema. This identifier can then be referenced elsewhere in the schema using the `$ref` keyword.

Reviewer: ```suggestion
The `$anchor` keyword is used to assign a unique identifier to a subschema within its schema resource. This identifier can then be referenced elsewhere using the `$ref` keyword.
```

Small clarifications on schema resources, and the fact that you can reference anchors even from outside the current schema

> * Its value must be a valid identifier starting with a letter and containing letters, digits, hyphens, underscores, colons, or periods.

Reviewer: Is this actually the case? I guess any valid URI fragment would do, but I might be wrong. i.e. `1234` or even something that looks like a JSON Pointer might in the theory do.

Can you cross check this with a few implementations and with the URI RFC?

Author: Okay, let me verify

Author: ![image](https://github.com/Intelligence-AI/learnjsonschema.com/assets/113545196/a62f51ec-5b7c-4862-b3d2-7134e701d8d1)
https://json-schema.org/draft/2019-09/json-schema-core#rfc.section.8.2.3

I tried with [hyperhump schema validator](https://json-schema.hyperjump.io/) which allows any string values (including special characters) without checking any format (which I think is not an expected behavior), whereas [json everything](https://json-everything.net/json-schema) strictly follows the above rule and only allows the `$achors` with  above-mentioned format.

Reviewer: Ah, great

Reviewer 2: If I remember correctly (and if it matters) this constraint didn't always exist. I think it was added in 2019-09. Sorry, for not looking it up myself.

* This keyword allows for the creation of plain name fragments, providing a flexible alternative to using JSON Pointer fragments which require knowledge of the schema's structure.

> * By using `$anchor`, schemas can be written with reusable components that aren't tied to specific structural locations.

Reviewer: This point seems similar to the previous one. Maybe merge them into one and simplify them?

> * These subschemas with named anchor are derived from the nearest `$id` or document base URI.

Reviewer: Maybe a more correct way of saying it would be:

```suggestion
* An anchor is resolved against the base URI of its schema resource
```
